test/functional/p2p_segwit.py

```diff
@@ -2066,6 +2069,89 @@ def received_wtxidrelay():
+            tx = CTransaction()
+            tx.vin.append(CTxIn(COutPoint(prev_txid, prev_index), b""))
+            tx.vout.append(CTxOut(amount, script_pubkey))
+            tx.rehash()
+            return tx
+
+        def sign_witness_input(tx, vin_index, script, amount, signing_key=key, multisig=False, low_s=True):
+            tx.wit.vtxinwit.append(CTxInWitness())
+            sighash = SegwitV0SignatureHash(script, tx, vin_index, SIGHASH_ALL, amount)
+            der_sig = signing_key.sign_ecdsa(sighash, low_s=low_s)
+            tx.wit.vtxinwit[vin_index].scriptWitness.stack.insert(0, der_sig + bytes([SIGHASH_ALL]))
+            if multisig:
+                # witness stack format is [b'', signature, script]
+                tx.wit.vtxinwit[vin_index].scriptWitness.stack.append(script)
+                tx.wit.vtxinwit[vin_index].scriptWitness.stack.insert(0, b'')
+            else:
+                # witness stack format is [signature, pubkey]
+                tx.wit.vtxinwit[vin_index].scriptWitness.stack.append(pubkey)
+            tx.rehash()
+
+        # Example 1: P2WPKH transaction with malleated signature
+        # 1. Create a funding transaction (P2WPKH output)
+        script_pubkey = key_to_p2wpkh_script(pubkey)
+        funding_p2wpkh_tx = create_tx(self.utxo[0].sha256, self.utxo[0].n, amount, script_pubkey)
+        block = self.build_next_block()
+        self.update_witness_block_with_transactions(block, [funding_p2wpkh_tx])
+        test_witness_block(self.nodes[0], self.test_node, block, accepted=True)
+
+        # 2. Create and sign a transaction spending from the P2WPKH output (bad high-S signature)
+        spending_tx = create_tx(funding_p2wpkh_tx.sha256, 0, amount - 1000, script_pubkey)
+        pubkey_hash = hash160(pubkey)
+        p2pkh_script = keyhash_to_p2pkh_script(pubkey_hash)
+        sign_witness_input(spending_tx, 0, p2pkh_script, amount, low_s=False)
+        prev_txid = spending_tx.hash
+        prev_wtxid = spending_tx.getwtxid()
+        # Submit transaction (should be rejected due to non-canonical signature)
+        test_transaction_acceptance(self.nodes[0], self.test_node, spending_tx, with_witness=True, accepted=False, reason='non-mandatory-script-verify-flag (Non-canonical signature: S value is unnecessarily high)')
+
+        # 3. Re-sign with low-S signature (standard)
+        spending_tx.wit.vtxinwit[0].scriptWitness.stack.clear()
+        sign_witness_input(spending_tx, 0, p2pkh_script, amount, low_s=True)
+
+        # 4. Verify txid matches but wtxid differs
+        assert_equal(prev_txid, spending_tx.hash)
+        assert_not_equal(prev_wtxid, spending_tx.getwtxid())
+        # Transaction should now be accepted
+        test_transaction_acceptance(self.nodes[1], self.test_node, spending_tx, with_witness=True, accepted=True)
+        self.generate(self.nodes[0], 1)
+
+        # Example 2: P2WSH 1-of-2 multisig transaction where each signature applied to tx results in different wtxid
```

Comment:
Isn't multisig kind of out-of-scope for creating a valid malleated transaction? I don't mind some extra tests, but those might make this PR more difficult to review. Maybe omit those if there is no review interest for some time.

Reply:
Now I see! Multisig is another way to create "same_txid, different_wtxid" even though it does not fiddle with the `s` value. And so, both transactions are relayable. Cool. Best would be to have a function that produces those two transactions. That would need the private keys. I guess better be separate from `create_malleated_version()`.